router/throttler/throttler.go

```diff
@@ -1,173 +1,156 @@
 package throttler
 
 import (
+	"context"
 	"fmt"
+	"sync"
 	"time"
 
+	"github.com/go-redis/redis/v8"
+
 	"github.com/rudderlabs/rudder-server/config"
-	"github.com/rudderlabs/rudder-server/router/throttler/ratelimiter"
-	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/internal/throttling"
+	"github.com/rudderlabs/rudder-server/services/stats"
 )
 
 const (
-	DESTINATION_LEVEL = "destination"
-	USER_LEVEL        = "user"
-	ALL_LEVELS        = "all"
+	throttlingAlgoTypeGoRate         = "gorate"
+	throttlingAlgoTypeGCRA           = "gcra"
+	throttlingAlgoTypeRedisGCRA      = "redis-gcra"
+	throttlingAlgoTypeRedisSortedSet = "redis-sorted-set"
 )
 
-// Throttler is an interface for throttling functions
-type Throttler interface {
-	CheckLimitReached(destID, userID string, currentTime time.Time) bool
-	Inc(destID, userID string, currentTime time.Time)
-	Dec(destID, userID string, count int64, currentTime time.Time, atLevel string)
-	IsEnabled() bool
-	IsUserLevelEnabled() bool
-	IsDestLevelEnabled() bool
-}
-
-type Limiter struct {
-	enabled     bool
-	eventLimit  int
-	timeWindow  time.Duration
-	ratelimiter *ratelimiter.RateLimiter
+type limiter interface {
+	// Allow returns true if the limit is not exceeded, false otherwise.
+	Allow(ctx context.Context, cost, rate, window int64, key string) (bool, func(context.Context) error, error)
 }
 
-type Settings struct {
-	limit               int
-	timeWindow          int64
-	userLevelLimit      int
-	userLevelTimeWindow int64
+type Factory struct {
+	Stats        stats.Stats
+	limiter      limiter
+	throttlers   map[string]*Throttler // map key is the destinationID
+	throttlersMu sync.Mutex
 }
 
-// HandleT is a Handle for event limiter
-type HandleT struct {
-	destinationName string
-	destLimiter     *Limiter
-	userLimiter     *Limiter
+// New constructs a new Throttler Factory
+func New() (*Factory, error) {
+	f := Factory{}
```

Comment:
Since we are using the New function, we could initialize its `throttlers` map here and skip the nil check in `Factory#Get`